balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,286 @@
+		bal := child.(balancer.Balancer)
+		if _, ok := endpointSet.Get(e); !ok {
+			bal.Close()
+			ba.children.Delete(e)
+		}
+	}
+	return ret
+}
+
+// ResolverError forwards the resolver error to all of the balancerAggregator's
+// children and sends a single synchronous update of the childStates at the end
+// of the ResolverError operation.
+func (ba *balancerAggregator) ResolverError(err error) {
+	ba.inhibitChildUpdates.Store(true)
+	defer func() {
+		ba.inhibitChildUpdates.Store(false)
+		ba.updateState()
+	}()
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.ResolverError(err)
+	}
+}
+
+func (ba *balancerAggregator) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// UpdateSubConnState is deprecated.
+}
+
+func (ba *balancerAggregator) Close() {
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.Close()
+	}
+}
+
+// updateState updates this component's state. It sends the aggregated state,
+// and a picker with round robin behavior with all the child states present if
+// needed.
+func (ba *balancerAggregator) updateState() {
+	if ba.inhibitChildUpdates.Load() {
+		return
+	}
+	readyPickers := make([]balancer.Picker, 0)
+	connectingPickers := make([]balancer.Picker, 0)
+	idlePickers := make([]balancer.Picker, 0)
+	transientFailurePickers := make([]balancer.Picker, 0)
+
+	childStates := make([]ChildState, 0, ba.children.Len())
+	ba.mu.Lock()
+	defer ba.mu.Unlock()
+	for _, child := range ba.children.Values() {
+		bw := child.(*balancerWrapper)
+		childState := bw.childState
+		childStates = append(childStates, childState)
+		childPicker := childState.State.Picker
+		switch childState.State.ConnectivityState {
+		case connectivity.Ready:
+			readyPickers = append(readyPickers, childPicker)
+		case connectivity.Connecting:
+			connectingPickers = append(connectingPickers, childPicker)
+		case connectivity.Idle:
+			idlePickers = append(idlePickers, childPicker)
+		case connectivity.TransientFailure:
+			transientFailurePickers = append(transientFailurePickers, childPicker)
+			// connectivity.Shutdown shouldn't appear.
+		}
+	}
+
+	// Construct the round robin picker based off the aggregated state. Whatever
+	// the aggregated state, use the pickers present that are currently in that
+	// state only.
+	var aggState connectivity.State
+	var pickers []balancer.Picker
+	if len(readyPickers) >= 1 {
+		aggState = connectivity.Ready
+		pickers = readyPickers
+	} else if len(connectingPickers) >= 1 {
+		aggState = connectivity.Connecting
+		pickers = connectingPickers
+	} else if len(idlePickers) >= 1 {
+		aggState = connectivity.Idle
+		pickers = idlePickers
+	} else if len(transientFailurePickers) >= 1 {
+		aggState = connectivity.TransientFailure
+		pickers = transientFailurePickers
+	} else {
+		aggState = connectivity.TransientFailure
+		pickers = append(pickers, base.NewErrPicker(errors.New("no children to pick from")))
```

Comment:
Remove the `append`?

Reply:
You could theoretically get more than one child in TF right? In the Java case that we discussed, it round robins over the error messages of the children in that case...with some intricacies as to whether the overall resolver error takes precedence or not.

Reply:
Isn't that what the above `pickers = transientFailurePickers` does?  This is with zero children, so there's only ever one, and pickers is set by exactly one of these if/else clauses.

Reply:
Oh whoops for some reason my eyes went to the else if above. I see what you're saying now. Switched.